File: rs/moq-secure/src/wire.rs
use crate::crypto::{aead_decrypt, aead_encrypt, sha256_digest};
use crate::error::MoqSecureError;
use crate::key_store::KeyStore;
use ed25519_dalek::{Signer, Verifier};

pub const MAGIC: [u8; 4] = *b"MOQS";
pub const VERSION: u8 = 1;

pub const SIG_SLOT_LEN: usize = 64;
pub const AEAD_TAG_LEN: usize = 16;
pub const PAD_LEN_FIELD_LEN: usize = 4;

// magic(4) | version(1) | key_id(1) | ctr(8) |
// n_signed(1) | sig_flag(1) | encrypted(1)
pub const FIXED_HEADER_LEN: usize = 4 + 1 + 1 + 8 + 1 + 1 + 1;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WireHeader {
    pub magic: [u8; 4],
    pub version: u8,
    pub key_id: u8,
    pub ctr: u64,
    pub n_signed: u8,
    pub sig_flag: u8,
    pub encrypted: u8,
}

impl WireHeader {
    pub fn encode(&self) -> Vec<u8> {
        let mut v = Vec::with_capacity(FIXED_HEADER_LEN);

        v.extend_from_slice(&self.magic);
        v.push(self.version);
        v.push(self.key_id);
        v.extend_from_slice(&self.ctr.to_be_bytes());
        v.push(self.n_signed);
        v.push(self.sig_flag);
        v.push(self.encrypted);

        v
    }

    pub fn aad(&self) -> Vec<u8> {
        self.encode()
    }

    pub fn validate(&self) -> Result<(), MoqSecureError> {
        if self.magic != MAGIC {
            return Err(MoqSecureError::InvalidMagic);
        }

        if self.version != VERSION {
            return Err(MoqSecureError::UnsupportedVersion(self.version));
        }

        if self.sig_flag != 0 && self.sig_flag != 1 {
            return Err(MoqSecureError::InvalidSigFlag(self.sig_flag));
        }

        if self.encrypted != 0 && self.encrypted != 1 {
            return Err(MoqSecureError::InvalidEncryptedFlag(self.encrypted));
        }

        if self.n_signed == 0 && self.sig_flag != 0 {
            return Err(MoqSecureError::SigningMismatch);
        }

        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct Frame {
    pub header: WireHeader,

    // For encrypted frames, payload is ciphertext without the AEAD tag.
    // For unencrypted frames, payload is:
    // padLen(4-byte big endian) || padding || plaintext.
    pub payload: Vec<u8>,

    pub tag: [u8; AEAD_TAG_LEN],
    pub signature: Option<[u8; SIG_SLOT_LEN]>,
}

impl Frame {
    pub fn parse(frame: &[u8]) -> Result<Self, MoqSecureError> {
        if frame.len() < FIXED_HEADER_LEN {
            return Err(MoqSecureError::TruncatedFrame);
        }

        let (h_bytes, rest0) = frame.split_at(FIXED_HEADER_LEN);
        let mut idx = 0;

        let mut magic = [0u8; 4];
        magic.copy_from_slice(&h_bytes[idx..idx + 4]);
        idx += 4;

        let version = h_bytes[idx];
        idx += 1;

        let key_id = h_bytes[idx];
        idx += 1;

        let mut ctr_bytes = [0u8; 8];
        ctr_bytes.copy_from_slice(&h_bytes[idx..idx + 8]);
        idx += 8;

        let ctr = u64::from_be_bytes(ctr_bytes);

        let n_signed = h_bytes[idx];
        idx += 1;

        let sig_flag = h_bytes[idx];
        idx += 1;

        let encrypted = h_bytes[idx];

        let header = WireHeader {
            magic,
            version,
            key_id,
            ctr,
            n_signed,
            sig_flag,
            encrypted,
        };

        header.validate()?;

        let sig_len = if header.sig_flag == 1 {
            SIG_SLOT_LEN
        } else {
            0
        };

        if rest0.len() < sig_len {
            return Err(MoqSecureError::TruncatedFrame);
        }

        let (payload_and_tag, signature_bytes) = if sig_len == 0 {
            (rest0, None)
        } else {
            let split_at = rest0.len() - SIG_SLOT_LEN;
            let (payload, signature) = rest0.split_at(split_at);
            (payload, Some(signature))
        };

        let signature = if let Some(sig_bytes) = signature_bytes {
            if sig_bytes.len() != SIG_SLOT_LEN {
                return Err(MoqSecureError::TruncatedFrame);
            }

            let mut sig = [0u8; SIG_SLOT_LEN];
            sig.copy_from_slice(sig_bytes);

            if sig == [0u8; SIG_SLOT_LEN] {
                return Err(MoqSecureError::InvalidSignature);
            }

            Some(sig)
        } else {
            None
        };

        if header.encrypted == 1 {
            if payload_and_tag.len() < AEAD_TAG_LEN {
                return Err(MoqSecureError::CiphertextTooShort);
            }

            let split_at = payload_and_tag.len() - AEAD_TAG_LEN;
            let ciphertext = &payload_and_tag[..split_at];
            let tag_bytes = &payload_and_tag[split_at..];

            let tag: [u8; AEAD_TAG_LEN] = tag_bytes
                .try_into()
                .map_err(|_| MoqSecureError::CiphertextTooShort)?;

            Ok(Self {
                header,
                payload: ciphertext.to_vec(),
                tag,
                signature,
            })
        } else {
            Ok(Self {
                header,
                payload: payload_and_tag.to_vec(),
                tag: [0u8; AEAD_TAG_LEN],
                signature,
            })
        }
    }

    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();

        out.extend_from_slice(&self.header.encode());

        if self.header.encrypted == 1 {
            out.extend_from_slice(&self.payload);
            out.extend_from_slice(&self.tag);
        } else {
            out.extend_from_slice(&self.payload);
        }

        if self.header.sig_flag == 1 {
            if let Some(signature) = self.signature {
                out.extend_from_slice(&signature);
            } else {
                out.extend_from_slice(&[0u8; SIG_SLOT_LEN]);
            }
        }

        out
    }

    pub fn aad_bytes(&self) -> Vec<u8> {
        self.header.aad()
    }

    pub fn digest_for_signature(&self) -> [u8; 32] {
        let header_bytes = self.header.encode();

        let mut data = Vec::with_capacity(
            header_bytes.len()
                + self.payload.len()
                + if self.header.encrypted == 1 {
                    AEAD_TAG_LEN
                } else {
                    0
                },
        );

        data.extend_from_slice(&header_bytes);
        data.extend_from_slice(&self.payload);

        if self.header.encrypted == 1 {
            data.extend_from_slice(&self.tag);
        }

        sha256_digest(&data)
    }

    pub fn decode_plaintext_with_key_store(
        &self,
        key_store: &dyn KeyStore,
        broadcaster_public_key: &ed25519_dalek::VerifyingKey,
        lease_remaining: &mut u8,
    ) -> Result<Vec<u8>, MoqSecureError> {
        if self.header.n_signed == 0 {
            if self.header.sig_flag != 0 || self.signature.is_some() {
                return Err(MoqSecureError::SigningMismatch);
            }
        } else if self.header.sig_flag == 1 {
            let signature_bytes = self
                .signature
                .ok_or(MoqSecureError::InvalidSignature)?;

            let signature =
                ed25519_dalek::Signature::from_bytes(&signature_bytes);

            let digest = self.digest_for_signature();

            broadcaster_public_key
                .verify(&digest, &signature)
                .map_err(|_| MoqSecureError::InvalidSignature)?;

            *lease_remaining = self.header.n_signed;
        } else {
            if *lease_remaining == 0 {
                return Err(MoqSecureError::InvalidSignature);
            }

            *lease_remaining -= 1;
        }

        let padded_plaintext = if self.header.encrypted == 1 {
            let key = key_store
                .aead_key(self.header.key_id)
                .ok_or(MoqSecureError::InvalidKeyId(self.header.key_id))?;

            aead_decrypt(
                key,
                self.header.key_id,
                self.header.ctr,
                &self.aad_bytes(),
                &self.payload,
                &self.tag,
            )?
        } else {
            self.payload.clone()
        };

        if padded_plaintext.len() < PAD_LEN_FIELD_LEN {
            return Err(MoqSecureError::InvalidPadLength);
        }

        let mut pad_len_bytes = [0u8; PAD_LEN_FIELD_LEN];
        pad_len_bytes.copy_from_slice(&padded_plaintext[..PAD_LEN_FIELD_LEN]);

        let pad_len = u32::from_be_bytes(pad_len_bytes) as usize;
        let content_start = PAD_LEN_FIELD_LEN
            .checked_add(pad_len)
            .ok_or(MoqSecureError::InvalidPadLength)?;

        if content_start > padded_plaintext.len() {
            return Err(MoqSecureError::InvalidPadLength);
        }

        Ok(padded_plaintext[content_start..].to_vec())
    }
}

pub fn encrypt_frame(
    key_store: &dyn KeyStore,
    broadcaster_private_key: &ed25519_dalek::SigningKey,
    key_id: u8,
    ctr: u64,
    n_signed: u8,
    maybe_sign: bool,
    encrypted: u8,
    pad_len: u32,
    plaintext: &[u8],
) -> Result<Frame, MoqSecureError> {
    if encrypted != 0 && encrypted != 1 {
        return Err(MoqSecureError::InvalidEncryptedFlag(encrypted));
    }

    let sig_flag = if n_signed != 0 && maybe_sign { 1 } else { 0 };

    let header = WireHeader {
        magic: MAGIC,
        version: VERSION,
        key_id,
        ctr,
        n_signed,
        sig_flag,
        encrypted,
    };

    header.validate()?;

    let pad_len_usize = pad_len as usize;
    let mut padded_plaintext = Vec::with_capacity(
        PAD_LEN_FIELD_LEN + pad_len_usize + plaintext.len(),
    );

    padded_plaintext.extend_from_slice(&pad_len.to_be_bytes());
    padded_plaintext.resize(PAD_LEN_FIELD_LEN + pad_len_usize, 0);
    padded_plaintext.extend_from_slice(plaintext);

    let frame_without_signature = if encrypted == 1 {
        let key = key_store
            .aead_key(key_id)
            .ok_or(MoqSecureError::InvalidKeyId(key_id))?;

        let (ciphertext, tag) = aead_encrypt(
            key,
            key_id,
            ctr,
            &header.aad(),
            &padded_plaintext,
        );

        Frame {
            header,
            payload: ciphertext,
            tag,
            signature: None,
        }
    } else {
        Frame {
            header,
            payload: padded_plaintext,
            tag: [0u8; AEAD_TAG_LEN],
            signature: None,
        }
    };

    if sig_flag == 1 {
        let digest = frame_without_signature.digest_for_signature();
        let signature = broadcaster_private_key.sign(&digest);

        Ok(Frame {
            signature: Some(signature.to_bytes()),
            ..frame_without_signature
        })
    } else {
        Ok(frame_without_signature)
    }
}

pub fn decrypt_frame(
    key_store: &dyn KeyStore,
    broadcaster_public_key: &ed25519_dalek::VerifyingKey,
    lease_remaining: &mut u8,
    frame_bytes: &[u8],
) -> Result<Vec<u8>, MoqSecureError> {
    let frame = Frame::parse(frame_bytes)?;

    frame.decode_plaintext_with_key_store(
        key_store,
        broadcaster_public_key,
        lease_remaining,
    )
}

File: js/src/wire.ts
import { createHash } from "node:crypto";
import * as ed25519 from "@noble/ed25519";

import {
  AEAD_TAG_LEN,
  FIXED_HEADER_LEN,
  MAGIC,
  SIG_SLOT_LEN,
  VERSION,
} from "./constants.js";
import { aeadDecrypt, aeadEncrypt, sha256Digest } from "./crypto.js";
import { MoqSecureError } from "./errors.js";
import type { KeyStore } from "./keys.js";
import {
  prependZeroPadding,
  removePadding,
} from "./padding.js";

ed25519.hashes.sha512 = (...messages: Uint8Array[]) => {
  const hash = createHash("sha512");

  for (const message of messages) {
    hash.update(message);
  }

  return new Uint8Array(hash.digest());
};

function equalBytes(a: Uint8Array, b: Uint8Array): boolean {
  return a.length === b.length &&
    a.every((value, index) => value === b[index]);
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const result = new Uint8Array(
    parts.reduce((length, part) => length + part.length, 0),
  );

  let offset = 0;

  for (const part of parts) {
    result.set(part, offset);
    offset += part.length;
  }

  return result;
}

function encodeUint32BE(value: number): Uint8Array {
  if (
    !Number.isSafeInteger(value) ||
    value < 0 ||
    value > 0xffffffff
  ) {
    throw new RangeError("padLen must be a uint32");
  }

  const result = new Uint8Array(4);
  new DataView(result.buffer).setUint32(0, value, false);
  return result;
}

export class WireHeader {
  constructor(
    public readonly magic: Uint8Array,
    public readonly version: number,
    public readonly keyId: number,
    public readonly ctr: bigint,
    public readonly nSigned: number,
    public readonly sigFlag: number,
    public readonly encrypted: number,
  ) {}

  encode(): Uint8Array {
    const result = new Uint8Array(FIXED_HEADER_LEN);
    let offset = 0;

    result.set(this.magic, offset);
    offset += 4;

    result[offset++] = this.version;
    result[offset++] = this.keyId;

    new DataView(result.buffer).setBigUint64(
      offset,
      this.ctr,
      false,
    );
    offset += 8;

    result[offset++] = this.nSigned;
    result[offset++] = this.sigFlag;
    result[offset++] = this.encrypted;

    return result;
  }

  aad(): Uint8Array {
    return this.encode();
  }

  validate(): void {
    if (!equalBytes(this.magic, MAGIC)) {
      throw MoqSecureError.invalidMagic();
    }

    if (this.version !== VERSION) {
      throw MoqSecureError.unsupportedVersion(this.version);
    }

    if (this.sigFlag !== 0 && this.sigFlag !== 1) {
      throw new MoqSecureError(
        "InvalidSigFlag",
        `sigFlag must be 0 or 1, got ${this.sigFlag}`,
      );
    }

    if (this.encrypted !== 0 && this.encrypted !== 1) {
      throw new MoqSecureError(
        "InvalidEncryptedFlag",
        `encrypted flag must be 0 or 1, got ${this.encrypted}`,
      );
    }

    if (this.nSigned === 0 && this.sigFlag !== 0) {
      throw new MoqSecureError(
        "SigningMismatch",
        "signing is disabled but sigFlag indicates a signature",
      );
    }
  }
}

export interface FrameInit {
  header: WireHeader;
  payload: Uint8Array;
  tag?: Uint8Array;
  signature?: Uint8Array;
}

export class Frame {
  readonly header: WireHeader;
  readonly payload: Uint8Array;
  readonly tag: Uint8Array;
  readonly signature?: Uint8Array;

  constructor(init: FrameInit) {
    this.header = init.header;
    this.payload = init.payload.slice();
    this.tag = (
      init.tag ?? new Uint8Array(AEAD_TAG_LEN)
    ).slice();
    this.signature = init.signature?.slice();
  }

  static parse(input: Uint8Array): Frame {
    if (input.length < FIXED_HEADER_LEN) {
      throw MoqSecureError.truncated();
    }

    const view = new DataView(
      input.buffer,
      input.byteOffset,
      input.byteLength,
    );

    const header = new WireHeader(
      input.slice(0, 4),
      input[4],
      input[5],
      view.getBigUint64(6, false),
      input[14],
      input[15],
      input[16],
    );

    header.validate();

    const signatureLength = header.sigFlag === 1
      ? SIG_SLOT_LEN
      : 0;

    if (input.length < FIXED_HEADER_LEN + signatureLength) {
      throw MoqSecureError.truncated();
    }

    const bodyEnd = input.length - signatureLength;
    const body = input.slice(FIXED_HEADER_LEN, bodyEnd);

    let signature: Uint8Array | undefined;

    if (signatureLength > 0) {
      signature = input.slice(bodyEnd);

      if (signature.length !== SIG_SLOT_LEN) {
        throw MoqSecureError.truncated();
      }

      if (signature.every((byte) => byte === 0)) {
        throw MoqSecureError.invalidSignature();
      }
    }

    if (header.encrypted === 1) {
      if (body.length < AEAD_TAG_LEN) {
        throw new MoqSecureError(
          "CiphertextTooShort",
          "ciphertext too short for AEAD tag",
        );
      }

      const ciphertextEnd = body.length - AEAD_TAG_LEN;

      return new Frame({
        header,
        payload: body.slice(0, ciphertextEnd),
        tag: body.slice(ciphertextEnd),
        signature,
      });
    }

    return new Frame({
      header,
      payload: body,
      signature,
    });
  }

  serialize(): Uint8Array {
    const body = this.header.encrypted === 1
      ? concat(this.payload, this.tag)
      : this.payload;

    const signature = this.header.sigFlag === 1
      ? this.signature ?? new Uint8Array(SIG_SLOT_LEN)
      : new Uint8Array();

    return concat(
      this.header.encode(),
      body,
      signature,
    );
  }

  aadBytes(): Uint8Array {
    return this.header.aad();
  }

  digestForSignature(): Uint8Array {
    const body = this.header.encrypted === 1
      ? concat(this.payload, this.tag)
      : this.payload;

    return sha256Digest(
      concat(this.header.encode(), body),
    );
  }

  async decodePlaintext(
    keyStore: KeyStore,
    broadcasterPublicKey: Uint8Array,
    lease: { remaining: number },
  ): Promise<Uint8Array> {
    if (this.header.nSigned === 0) {
      if (this.header.sigFlag !== 0 || this.signature) {
        throw new MoqSecureError(
          "SigningMismatch",
          "signing is disabled but a signature is present",
        );
      }
    } else if (this.header.sigFlag === 1) {
      if (!this.signature) {
        throw MoqSecureError.invalidSignature();
      }

      const valid = await ed25519.verify(
        this.signature,
        this.digestForSignature(),
        broadcasterPublicKey,
      );

      if (!valid) {
        throw MoqSecureError.invalidSignature();
      }

      lease.remaining = this.header.nSigned;
    } else {
      if (lease.remaining === 0) {
        throw MoqSecureError.invalidSignature();
      }

      lease.remaining--;
    }

    let paddedPlaintext: Uint8Array;

    if (this.header.encrypted === 1) {
      const key = keyStore.aeadKey(this.header.keyId);

      if (!key) {
        throw new MoqSecureError(
          "InvalidKeyId",
          `unknown or not-loaded key_id: ${this.header.keyId}`,
          this.header.keyId,
        );
      }

      // The encrypted plaintext is:
      //
      //   uint32be(padLen) || zero padding || plaintext
      //
      // The pad_len field is part of the encrypted plaintext.
      paddedPlaintext = aeadDecrypt(
        key,
        this.header.keyId,
        this.header.ctr,
        this.aadBytes(),
        this.payload,
        this.tag,
      );
    } else {
      // Cleartext payload has the same layout:
      //
      //   uint32be(padLen) || zero padding || plaintext
      paddedPlaintext = this.payload;
    }

    try {
      return removePadding(paddedPlaintext);
    } catch {
      throw new MoqSecureError(
        "InvalidPadLength",
        "invalid padding length",
      );
    }
  }
}

export async function encryptFrame(
  keyStore: KeyStore,
  broadcasterPrivateKey: Uint8Array,
  keyId: number,
  ctr: bigint,
  nSigned: number,
  maybeSign: boolean,
  encrypted: number,
  padLen: number,
  plaintext: Uint8Array,
): Promise<Frame> {
  if (encrypted !== 0 && encrypted !== 1) {
    throw new MoqSecureError(
      "InvalidEncryptedFlag",
      `encrypted flag must be 0 or 1, got ${encrypted}`,
    );
  }

  const sigFlag = nSigned !== 0 && maybeSign ? 1 : 0;

  const header = new WireHeader(
    MAGIC,
    VERSION,
    keyId,
    ctr,
    nSigned,
    sigFlag,
    encrypted,
  );

  header.validate();

  // The plaintext representation is:
  //
  //   uint32be(padLen) || zero padding || plaintext
  //
  // For encrypted frames, the complete value is encrypted.
  const paddedPlaintext = concat(
    encodeUint32BE(padLen),
    prependZeroPadding(plaintext, padLen),
  );

  let payload: Uint8Array;
  let tag = new Uint8Array(AEAD_TAG_LEN);

  if (encrypted === 1) {
    const key = keyStore.aeadKey(keyId);

    if (!key) {
      throw new MoqSecureError(
        "InvalidKeyId",
        `unknown or not-loaded key_id: ${keyId}`,
        keyId,
      );
    }

    const encryptedResult = aeadEncrypt(
      key,
      keyId,
      ctr,
      header.aad(),
      paddedPlaintext,
    );

    // The tag is stored separately by Frame.
    payload = encryptedResult.ciphertext;
    tag = encryptedResult.tag;
  } else {
    // Cleartext payload contains the same pad_len-prefixed layout.
    payload = paddedPlaintext;
  }

  const unsignedFrame = new Frame({
    header,
    payload,
    tag,
  });

  if (sigFlag === 0) {
    return unsignedFrame;
  }

  const signature = await ed25519.sign(
    unsignedFrame.digestForSignature(),
    broadcasterPrivateKey,
  );

  return new Frame({
    header,
    payload,
    tag,
    signature,
  });
}

export async function decryptFrame(
  keyStore: KeyStore,
  broadcasterPublicKey: Uint8Array,
  lease: { remaining: number },
  frameBytes: Uint8Array,
): Promise<Uint8Array> {
  const frame = Frame.parse(frameBytes);

  return frame.decodePlaintext(
    keyStore,
    broadcasterPublicKey,
    lease,
  );
}
